kotlinx-coroutines-core/common/src/flow/operators/Delay.kt

```diff
@@ -348,3 +348,92 @@ internal fun CoroutineScope.fixedPeriodTicker(delayMillis: Long, initialDelayMil
 @ExperimentalTime
 @FlowPreview
 public fun <T> Flow<T>.sample(period: Duration): Flow<T> = sample(period.toDelayMillis())
+
+/**
+ * Returns a flow that will emit a [TimeoutCancellationException] if the upstream doesn't emit an item within the given time.
+ *
+ * Example:
+ *
+ * ```kotlin
+ * flow {
+ *     emit(1)
+ *     delay(100)
+ *     emit(2)
+ *     delay(100)
+ *     emit(3)
+ *     delay(1000)
+ *     emit(4)
+ * }.timeout(100.milliseconds).catch {
+ *     emit(-1) // Item to emit on timeout
+ * }.onEach {
+ *     delay(300) // This will not cause a timeout
+ * }
+ * ```
+ * <!--- KNIT example-timeout-duration-01.kt -->
+ *
+ * produces the following emissions
+ *
+ * ```text
+ * 1, 2, 3, -1
+ * ```
+ * <!--- TEST -->
+ *
+ * Note that delaying on the downstream doesn't trigger the timeout.
+ *
+ * @param timeout period. If non-positive, the flow is timed out immediately
+ */
+@FlowPreview
+public fun <T> Flow<T>.timeout(
+    timeout: Duration
+): Flow<T> = timeoutInternal(timeout.toDelayMillis())
+
+private fun <T> Flow<T>.timeoutInternal(
+    timeoutMillis: Long
+): Flow<T> = scopedFlow { downStream ->
+    if (timeoutMillis <= 0L) throw TimeoutCancellationException("Timed out immediately")
+
+    // Produce the values using the default (rendezvous) channel
+    // Similar to [debounceInternal]
+    val values = produce<Any?> {
+        var timeoutJob = launch { // Emits timeout unless cancelled
+            delay(timeoutMillis)
+            send(TIMEOUT)
+        }
+        try {
+            collect {
+                timeoutJob.cancel() // Upstream emitted, so cancel the job
+
+                send(it ?: NULL)
+
+                // We reset the job here. The reason being is that the `flow.emit()` suspends, which in turn suspends `send()`.
+                // We only want to measure a timeout if the producer took longer than `timeoutMillis`, not producer + consumer
+                timeoutJob = launch {
+                    delay(timeoutMillis)
+                    send(TIMEOUT)
+                }
+            }
+        } finally {
+            timeoutJob.cancel()
+            send(DONE) // Special signal to let flow end
+        }
+    }
+
+    // Await for values from our producer now
+    for (value in values) {
+        if (value !== DONE) {
+            if (value === TIMEOUT) {
+                throw TimeoutCancellationException("Timed out waiting for $timeoutMillis ms")
```

Comment:
Probably worth using `duration` everywhere and avoid using `Long`-based MS everywhere: `onTimeout` already supports it and exception message will be more friendly